acvm-repo/acvm/src/compiler/optimizers/witness_rebinding.rs

```diff
@@ -0,0 +1,605 @@
+use std::collections::{BTreeSet, HashMap};
+
+use crate::compiler::optimizers::GeneralOptimizer;
+use acir::{
+    circuit::{
+        brillig::{BrilligInputs, BrilligOutputs},
+        directives::Directive,
+        opcodes::{BlackBoxFuncCall, ConstantOrWitnessEnum, FunctionInput, MemOp},
+        Circuit, Opcode,
+    },
+    native_types::{Expression, Witness},
+    AcirField,
+};
+
+/// `WitnessRebindingOptimizer` will remove redundant range constraints.
+///
+/// # Example
+///
+/// Suppose we had the following circuit where witness 0 is an input and
+/// witness 2 is a return value:
+///
+/// ```
+/// EXPR [ (-1, _0) (1, _1) 0 ]
+/// EXPR [ (-1, _1) (1, _2) 0 ]
+/// ```
+///
+/// While it's not possible to determine the values of any of the witnesses, it's clear that they will all be equal.
+/// As witness 1 is internal to the circuit we can then remove it and replace any references to it with witness 0.
+///
+/// ```
+/// EXPR [ (-1, _0) (1, _2) 0 ]
+/// ```
+///
+/// This optimization pass looks for [`Opcode::AssertZero`] opcodes of this form and replaces references and replaces
+/// all references to one of these witnesses with the other to apply this equality in all other opcodes.
```

Comment:
> and replaces references and replaces
> all references to one of these witnesses with the other to apply this equality in all other opcodes.

-> `and applies this equality in all other opcodes by replacing all references to one of these witnesses with the other.`